c_glib/arrow-glib/file-system.cpp

```diff
@@ -0,0 +1,1092 @@
+ * @path: The path of the output stream.
+ * @error: (nullable): Return location for a #GError or %NULL.
+ *
+ * Open an output stream for sequential writing.
+ * If the target already exists, the existing data is truncated.
+ *
+ * Returns: (nullable) (transfer full): A newly created #GArrowOutputStream
+ *   for appending.
+ *
+ * Since: 1.0.0
+ */
+GArrowOutputStream *
+garrow_file_system_open_output_stream(GArrowFileSystem *file_system,
+                                      const gchar *path,
+                                      GError **error)
+{
+  auto arrow_file_system = garrow_file_system_get_raw(file_system);
+  auto arrow_output_stream = arrow_file_system->OpenOutputStream(path);
+  if (garrow::check(error, arrow_output_stream, "[filesystem][open-append-stream]")) {
+    return garrow_output_stream_new_raw(&(arrow_output_stream.ValueOrDie()));
+  } else {
+    return NULL;
+  }
+}
+
+/**
+ * garrow_file_system_open_append_stream:
+ * @file_system: A #GArrowFileSystem.
+ * @path: The path of the output stream.
+ * @error: (nullable): Return location for a #GError or %NULL.
+ *
+ * Open an output stream for appending.
+ * If the target doesn't exist, a new empty file is created.
+ *
+ * Returns: (nullable) (transfer full): A newly created #GArrowOutputStream
+ *   for appending.
+ *
+ * Since: 1.0.0
+ */
+GArrowOutputStream *
+garrow_file_system_open_append_stream(GArrowFileSystem *file_system,
+                                      const gchar *path,
+                                      GError **error)
+{
+  auto arrow_file_system = garrow_file_system_get_raw(file_system);
+  auto arrow_output_stream = arrow_file_system->OpenAppendStream(path);
+  if (garrow::check(error, arrow_output_stream, "[filesystem][open-append-stream]")) {
+    return garrow_output_stream_new_raw(&(arrow_output_stream.ValueOrDie()));
+  } else {
+    return NULL;
+  }
+}
+
+/* arrow::fs::SubTreeFileSystem */
+
+G_DEFINE_TYPE(GArrowSubTreeFileSystem,
+              garrow_sub_tree_file_system,
+              GARROW_TYPE_FILE_SYSTEM)
+
+static void
+garrow_sub_tree_file_system_init(GArrowSubTreeFileSystem *file_system)
+{
+}
+
+static void
+garrow_sub_tree_file_system_class_init(GArrowSubTreeFileSystemClass *klass)
+{
+}
+
+/**
+ * garrow_sub_tree_file_system_new:
+ * @base_path: A base path of the sub tree file system.
+ * @base_file_system: A #GArrowFileSystem as the base file system.
+ *
+ * Returns: (transfer full): A newly created #GArrowSubTreeFileSystem.
+ *
+ * Since: 1.0.0
+ */
+GArrowSubTreeFileSystem *
+garrow_sub_tree_file_system_new(const gchar *base_path,
+                                GArrowFileSystem *base_file_system)
+{
+  auto arrow_base_file_system = garrow_file_system_get_raw(base_file_system);
+  auto arrow_sub_tree_file_system =
+    std::make_shared<arrow::fs::SubTreeFileSystem>(base_path, arrow_base_file_system);
+  std::shared_ptr<arrow::fs::FileSystem> arrow_file_system = arrow_sub_tree_file_system;
```

Comment:
Can we use `auto arrow_file_system = std::static_pointer_cast<arrow::fs::FileSystem>(arrow_sub_tree_file_system)`?